### PR TITLE
HERA-434 Scaffolds [gamification]

### PR DESCRIPTION
**Pull request by UvgenGen**
[HERA-434](https://youtrack.raccoongang.com/issue/HERA-434) Scaffolds [gamification]

### CODE REVIEW

File: hera_xblocks/question/question/static/js/src/question.js [hunk 1]
```diff
@@ -2,6 +2,32 @@ function QuestionXBlock(runtime, element, init_args) {
     var submithHandlerUrl = runtime.handlerUrl(element, 'submit');
     var skipHandlerUrl = runtime.handlerUrl(element, 'skip');
     var renderHtmlHandlerUrl = runtime.handlerUrl(element, 'render_html');
+    var scaffoldPaymentHandlerUrl = runtime.handlerUrl(element, 'scaffold_payment');
+    var scaffolds = init_args.scaffolds
+
+    function scaffoldPayment(scaffoldName){
+        var status = true
```
Comment (sendr) [inline hera_xblocks/question/question/static/js/src/question.js]: what is this var about?

Comment (sendr) [reply hera_xblocks/question/question/static/js/src/question.js]: where is semicolon?

File: lms/djangoapps/courseware/views/index.py [hunk 2]
```diff
@@ -100,6 +102,35 @@ def get(self, request, course_id, chapter=None, section=None, position=None):
...
+
+        if not user_lesson_coins:
+            c_user_lesson_coins().update(
+                {
+                    BLOCK_ID: section,
+                    USER: request.user.id
+                },
+                {
+                    '$set': {
+                        'coins': Scaffold.get_scaffold().starting_coins,
```
Comment (sendr) [inline lms/djangoapps/courseware/views/index.py]: Rename please `Scaffold` and `get_scaffold`

File: lms/djangoapps/courseware/views/index.py [hunk 3]
```diff
@@ -100,6 +102,35 @@ def get(self, request, course_id, chapter=None, section=None, position=None):
...
+                    }
+                },
+                upsert=True
+            )
+            user_lesson_coins = c_user_lesson_coins().find_one(
```
Comment (sendr) [inline lms/djangoapps/courseware/views/index.py]: please avoid this request

File: lms/djangoapps/courseware/views/index.py [hunk 4]
```diff
@@ -362,8 +393,12 @@ def _create_courseware_context(self, request):
         """
         course_url_name = default_course_url_name(self.course.id)
         course_url = reverse(course_url_name, kwargs={'course_id': unicode(self.course.id)})
+        scaffold = Scaffold.get_scaffold()
```
Comment (sendr) [inline lms/djangoapps/courseware/views/index.py]: should be renamed

File: lms/djangoapps/courseware/views/index.py [hunk 5]
```diff
@@ -362,8 +393,12 @@ def _create_courseware_context(self, request):
...
 
         courseware_context = {
+            'is_show_coins': True,
+            'user_coins': self.user_coins,
+            'coins_img_url': scaffold.coin_icon.url if scaffold.coin_icon else None,
```
Comment (sendr) [inline lms/djangoapps/courseware/views/index.py]: is it worth making the method `get_coin_img_url` or something like that?

File: lms/djangoapps/hera/admin.py [hunk 6]
```diff
@@ -97,6 +99,28 @@ def has_add_permission(self, request):
         return not self.model.objects.exists()
 
 
+class ScaffoldForm(ModelForm):
+    class Meta:
+        model = Scaffold
+        fields = '__all__'
+        widgets = {
+            'rephrase_color': TextInput(attrs={'type': 'color'}),
```
Comment (sendr) [inline lms/djangoapps/hera/admin.py]: should be checked whether it works in all browsers.

File: lms/djangoapps/hera/models.py [hunk 7]
```diff
@@ -322,6 +322,21 @@ def user_dashboard_img_urls(cls):
         return {}
 
 
+class Scaffold(models.Model):
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]: should be renamed into `ScaffoldsSettings`

File: lms/djangoapps/hera/models.py [hunk 8]
```diff
@@ -322,6 +322,21 @@ def user_dashboard_img_urls(cls):
...
+    rephrase_cost = models.IntegerField(default=10)
+    rephrase_color = models.CharField(max_length=7, default="#257bba")
+    break_it_down_cost = models.IntegerField(default=20)
+    break_it_down_color = models.CharField(max_length=7, default="#674172")
+    teach_me_cost = models.IntegerField(default=30)
+    teach_me_color = models.CharField(max_length=7, default="#ec8b22")
+    starting_coins = models.IntegerField(default=300)
+
+    @classmethod
+    def get_scaffold(cls):
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]: please rename it

File: lms/djangoapps/hera/utils.py [hunk 9]
```diff
@@ -41,3 +45,41 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
                     break
 
     return {'lesson_title': lesson_title, 'points': points, 'medal': get_medal(points)}
+
+
+def recalculate_coints(course_id, block_id, user_id, cost=0):
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]: please add a docstring

File: lms/djangoapps/hera/utils.py [hunk 10]
```diff
@@ -41,3 +45,41 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
...
+                for block in unit.get_children():
+                    if block.scope_ids.usage_id.block_id == block_id:
+                        subsection_id = subsection.scope_ids.usage_id.block_id
+                        break
+
+    user_lesson_coins = c_user_lesson_coins().find_one({
+        BLOCK_ID: subsection_id,
+        USER: user_id,
+    })
+    if user_lesson_coins and cost <= user_lesson_coins.get('coins'):
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]: variable? `user_lesson_coins.get('coins')`

File: lms/djangoapps/hera/admin.py (hunk 6 above)
Comment (UvgenGen) [reply lms/djangoapps/hera/admin.py]: works in safari, chrome and firefox.

File: lms/djangoapps/hera/utils.py (hunk 9 above)
Comment (UvgenGen) [reply lms/djangoapps/hera/utils.py]: Added

File: hera_xblocks/lesson_summary/lesson_summary/static/html/lesson_summary.html [hunk 11]
```diff
@@ -5,30 +5,32 @@
...
+                            <div class="bubble-columns__icon default-icon__medal">
                             </div>
```
Comment (sendr) [inline hera_xblocks/lesson_summary/lesson_summary/static/html/lesson_summary.html]:
```suggestion
                            <div class="bubble-columns__icon default-icon__medal"></div>
```

File: hera_xblocks/question/question/question.py [hunk 12]
```diff
@@ -2,16 +2,20 @@
 import pkg_resources
 from web_fragments.fragment import Fragment
 from xblock.core import XBlock
-from xblock.scorable import ScorableXBlockMixin
-from xblock.fields import JSONField, Scope, Integer, String
+from xblock.fields import JSONField, Scope, Integer, String, Boolean
 from xblockutils.resources import ResourceLoader
 from xblockutils.studio_editable import StudioEditableXBlockMixin
 
+from hera.utils import recalculate_coints, get_scaffold_settings
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]: `get_scaffolds_settings`

File: hera_xblocks/question/question/question.py [hunk 13]
```diff
@@ -142,6 +172,49 @@ def render_html(self, data, sufix=''):
     def get_data(self, somedata, sufix=''):
         return self.data
 
+    @XBlock.json_handler
+    def scaffold_payment(self, data, sufix=''):
+        scaffold_name = data.get("scaffold_name")
+        scaffold_settings = get_scaffold_settings()
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]:
```suggestion
        scaffold_settings = get_scaffold_settings()
        scaffold_name_mapping = {
            REPHRASE: scaffold_settings.rephrase_cost,
            BREAK_IT_DOWN: scaffold_settings.break_it_down_cost,
            TEACH_ME: scaffold_settings.teach_me_cost
        }
        coins = recalculate_coints(
                 str(self.course_id),
                 self.location.block_id,
                 self.scope_ids.user_id,
                 scaffold_name_mapping[scaffold_name]
             )
```

File: hera_xblocks/question/question/question.py [hunk 14]
```diff
@@ -142,6 +172,49 @@ def render_html(self, data, sufix=''):
...
+        if scaffold_name == REPHRASE:
+            coins = recalculate_coints(
+                str(self.course_id),
+                self.location.block_id,
+                self.scope_ids.user_id,
+                scaffold_settings.rephrase_cost
+            )
+            self.rephrase_paid = True if not coins is None else False
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]:
```suggestion
            scaffold_paid_mapping = {
                REPHRASE: 'rephrase_paid',
                BREAK_IT_DOWN: 'break_it_down_paid',
                TEACH_ME: 'teach_me_paid'
            }
            if coins is not None:
                setattr(self, scaffold_paid_mapping[scaffold_name], coins)
            return {
                'coins': coins,
                'scaffold_paid': getattr(self, scaffold_paid_mapping[scaffold_name])
            }
```

File: hera_xblocks/question/question/static/html/question.html [hunk 15]
```diff
@@ -103,50 +103,62 @@
...
+            <div id="${'{}-{}'.format(teach_me_name, block_id)}" class="teach-me-content" hidden>
+                ${scaffolds[teach_me_name].get('data').get('content')}
             </div>
         </div>
-        <div class="scaffolds hidden">
+        <div class="scaffolds ${'' if is_submission_allowed and submission_counter > 0 else 'hidden'}">
             <%
-            break_down_img_urls = ' '.join(break_down.get('imgUrls', []))
-            teach_me_img_urls = ' '.join(teach_me.get('imgUrls', []))
+            break_it_down_img_urls = ' '.join(scaffolds[break_it_down_name].get('data').get('imgUrls', []))
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]:
```suggestion
            break_it_down_img_urls = ' '.join(scaffolds[break_it_down_name]['data'].get('imgUrls', []))
```

File: hera_xblocks/question/question/static/html/question.html [hunk 16]
```diff
@@ -103,50 +103,62 @@
...
                     class="scaffold scaffold__btn is-break js-scaffold-button"
-                    data-img-urls="${break_down_img_urls}"
-                    data-scaffold-name="break-down"
+                    data-img-urls="${break_it_down_img_urls}"
+                    data-scaffold-name="${break_it_down_name}"
                     data-scaffold-class-name="is-break">
-                    <span class="scaffold__price">
-                        -20
+                    <span class="scaffold__price ${break_it_down_name}">
+                        %if scaffolds[break_it_down_name].get('paid'):
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]:
```suggestion
                        %if scaffolds[break_it_down_name]['paid']:
```

File: hera_xblocks/question/question/static/js/src/question.js [hunk 17]
```diff
@@ -2,9 +2,35 @@ function QuestionXBlock(runtime, element, init_args) {
     var submithHandlerUrl = runtime.handlerUrl(element, 'submit');
     var skipHandlerUrl = runtime.handlerUrl(element, 'skip');
     var renderHtmlHandlerUrl = runtime.handlerUrl(element, 'render_html');
+    var scaffoldPaymentHandlerUrl = runtime.handlerUrl(element, 'scaffold_payment');
+    var scaffolds = init_args.scaffolds;
+
+    function scaffoldPayment(scaffoldName){
+        var isScaffoldsPaid = true;
```
Comment (sendr) [inline hera_xblocks/question/question/static/js/src/question.js]:
```suggestion
        var isScaffoldPaid = true;
```

File: lms/djangoapps/courseware/views/index.py [hunk 18]
```diff
@@ -362,8 +389,11 @@ def _create_courseware_context(self, request):
         """
         course_url_name = default_course_url_name(self.course.id)
         course_url = reverse(course_url_name, kwargs={'course_id': unicode(self.course.id)})
-
+        scaffold_settings = ScaffoldSettings.get()
         courseware_context = {
+            'is_show_coins': True,
```
Comment (sendr) [inline lms/djangoapps/courseware/views/index.py]: `show_coins`

File: lms/djangoapps/hera/models.py [hunk 19]
```diff
@@ -322,6 +322,27 @@ def user_dashboard_img_urls(cls):
...
+    rephrase_cost = models.IntegerField(default=10)
+    rephrase_color = models.CharField(max_length=7, default="#257bba")
+    break_it_down_cost = models.IntegerField(default=20)
+    break_it_down_color = models.CharField(max_length=7, default="#674172")
+    teach_me_cost = models.IntegerField(default=30)
+    teach_me_color = models.CharField(max_length=7, default="#ec8b22")
+    starting_coins = models.IntegerField(default=300)
+
+    class Meta(object):
+        verbose_name_plural = "Scaffold settings"
```
Comment (sendr) [inline lms/djangoapps/hera/models.py]:
```suggestion
        verbose_name_plural = "Scaffolds settings"
```

File: lms/djangoapps/hera/utils.py [hunk 20]
```diff
@@ -41,3 +45,45 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
...
+                for block in unit.get_children():
+                    if block.scope_ids.usage_id.block_id == block_id:
+                        subsection_id = subsection.scope_ids.usage_id.block_id
+                        break
+
+    user_lesson_coins = c_user_lesson_coins().find_one({
+        BLOCK_ID: subsection_id,
+        USER: user_id,
+    })
+    user_coins = user_lesson_coins.get('coins')
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]:
```suggestion
    user_coins = user_lesson_coins.get('coins', 0)
```

File: lms/djangoapps/hera/utils.py [hunk 21]
```diff
@@ -41,3 +45,45 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
...
+    if not user_coins is None and cost <= user_coins:
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]:
```suggestion
    if cost <= user_coins:
```

File: lms/djangoapps/hera/utils.py [hunk 22]
```diff
@@ -41,3 +45,45 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
                     break
 
     return {'lesson_title': lesson_title, 'points': points, 'medal': get_medal(points)}
+
+
+def recalculate_coints(course_id, block_id, user_id, cost=0):
+    """
+    Return difference between the cost and user coins.
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]: please rephrase the docstring

File: lms/djangoapps/hera/utils.py [hunk 23]
```diff
@@ -41,3 +45,46 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
                     break
 
     return {'lesson_title': lesson_title, 'points': points, 'medal': get_medal(points)}
+
+
+def recalculate_coints(course_id, block_id, user_id, cost=0):
+    """
+    Return difference between the user coins and cost.
+    If cost more then user coins cost - return None.
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]: pllease corerct the typo

File: lms/djangoapps/hera/utils.py (hunk 9 above)
Comment (sendr) [reply lms/djangoapps/hera/utils.py]:
```suggestion
def recalculate_coins(course_id, block_id, user_id, cost=0):
```

File: lms/djangoapps/hera/utils.py [hunk 24]
```diff
@@ -41,3 +45,46 @@ def get_lesson_summary_xblock_context(user, course_key, current_unit):
                     break
 
     return {'lesson_title': lesson_title, 'points': points, 'medal': get_medal(points)}
+
+
+def recalculate_coints(course_id, block_id, user_id, cost=0):
+    """
+    Return difference between the user coins and cost.
+    If cost more then user coins - return None.
```
Comment (sendr) [inline lms/djangoapps/hera/utils.py]: If user has less coins than given cost - return None.

**Pull request merged**